clippy_lints/src/operators/float_cmp.rs

```diff
@@ -61,54 +101,105 @@ pub(crate) fn check<'tcx>(
+        ExprKind::Path(_) | ExprKind::Lit(_) => true,
+        ExprKind::Field(e, _) | ExprKind::Cast(e, _) | ExprKind::Repeat(e, _) => is_pure_expr(cx, e),
+        ExprKind::Tup(args) => args.iter().all(|arg| is_pure_expr(cx, arg)),
+        ExprKind::Struct(_, fields, base) => {
+            base.map_or(true, |base| is_pure_expr(cx, base)) && fields.iter().all(|f| is_pure_expr(cx, f.expr))
+        },
+
+        // Since rust doesn't actually have the concept of a pure function we
+        // have to guess whether it's likely pure from the signature of the
+        // function.
+        ExprKind::Unary(_, e) => is_pure_arg_ty(cx, cx.typeck_results().expr_ty_adjusted(e)) && is_pure_expr(cx, e),
+        ExprKind::Binary(_, x, y) | ExprKind::Index(x, y, _) => {
+            is_pure_arg_ty(cx, cx.typeck_results().expr_ty_adjusted(x))
+                && is_pure_arg_ty(cx, cx.typeck_results().expr_ty_adjusted(y))
+                && is_pure_expr(cx, x)
+                && is_pure_expr(cx, y)
+        },
+        ExprKind::MethodCall(_, recv, args, _) => {
+            is_pure_arg_ty(cx, cx.typeck_results().expr_ty_adjusted(recv))
+                && is_pure_expr(cx, recv)
+                && cx
+                    .typeck_results()
+                    .type_dependent_def_id(e.hir_id)
+                    .is_some_and(|did| matches!(cx.tcx.fn_sig(did).skip_binder().skip_binder().safety, Safety::Safe))
+                && args
+                    .iter()
+                    .all(|arg| is_pure_arg_ty(cx, cx.typeck_results().expr_ty_adjusted(arg)) && is_pure_expr(cx, arg))
+        },
+        ExprKind::Call(f, args @ [_, ..]) => {
+            is_pure_expr(cx, f)
+                && is_pure_fn_ty(cx, cx.typeck_results().expr_ty_adjusted(f))
+                && args
+                    .iter()
+                    .all(|arg| is_pure_arg_ty(cx, cx.typeck_results().expr_ty_adjusted(arg)) && is_pure_expr(cx, arg))
+        },
+
+        _ => false,
     }
+}
+
+fn is_pure_fn_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    let sig = match *ty.peel_refs().kind() {
+        ty::FnDef(did, _) => cx.tcx.fn_sig(did).skip_binder(),
+        ty::FnPtr(sig) => sig,
+        ty::Closure(_, args) => {
+            return args.as_closure().upvar_tys().iter().all(|ty| is_pure_arg_ty(cx, ty));
+        },
+        _ => return false,
+    };
+    matches!(sig.skip_binder().safety, Safety::Safe)
+}
 
+fn is_pure_arg_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    !ty.is_mutable_ptr()
+        && ty.is_copy_modulo_regions(cx.tcx, cx.param_env)
+        && (ty.peel_refs().is_freeze(cx.tcx, cx.param_env)
+            || !ty.has_type_flags(TypeFlags::HAS_FREE_REGIONS | TypeFlags::HAS_RE_ERASED | TypeFlags::HAS_RE_BOUND))
+}
+
+fn contains_expr<'tcx>(cx: &LateContext<'tcx>, corpus: &'tcx Expr<'tcx>, e: &'tcx Expr<'tcx>) -> bool {
+    for_each_expr_without_closures(corpus, |corpus| {
+        if SpanlessEq::new(cx).eq_expr(corpus, e) {
```

Comment:
I have the feeling calling this in a closure can quickly explode...

@blyxyas You have a perf setup for Clippy right? Could you benchmark this PR?

Reply:
Currently benchmarking it!

Reply:
Worst case is O(n*m), but that requires nearly identical expressions on both sides which also contains a chain of nearly identical expressions as it's children. Something like `1 + 1 + 1 + 1 + 1 + 1 + x == 1 + 1 + 1 + 1 + 1 + x` would fit this pattern, but `1 + 2 + 3 + 4 + x == 1 + 2 + 3 + 4 + 5 + x` would not.

Basically it's a very degenerate case, and the expressions would still need to be quite large in practice. Most cases are very close to linear in complexity.

Reply:
@blyxyas Have there been any significant changes?

Reply:
Instructions \<most relevant\>

# Summary
| | Range | Mean | Count |
|:---:|:---:|:---:|:---:|
| Regressions | 0.63%, 2.85% | 1.68% | 3 |
| Improvements | -0.64%, -0.31% | -0.44% | 6 |
| All | -0.64%, 2.85% | 0.27% | 9 |

# Primary benchmarks
| Benchmark | Profile | Scenario | % Change | Significance Factor |
|:---:|:---:|:---:|:---:|:---:|
| cargo-0.60.0 | clippy | full | -0.34% | 1.71x
| webrender-2022 | clippy | full | -0.34% | 1.69x

# Secondary benchmarks
| Benchmark | Profile | Scenario | % Change | Significance Factor |
|:---:|:---:|:---:|:---:|:---:|
| coercions | clippy | incr-full | 2.85% | 14.24x
| coercions | clippy | incr-patched: println | 1.57% | 7.87x
| deep-vector | clippy | incr-patched: add vec item | -0.64% | 3.22x
| issue-46449 | clippy | full | 0.63% | 3.15x
| coercions | clippy | incr-patched: add static arr item | -0.59% | 2.94x
| coercions | clippy | full | -0.39% | 1.96x
| token-stream-stress | clippy | full | -0.31% | 1.57x

<details>
<summary>Cycles (less accurate)</summary>
# Summary
| | Range | Mean | Count |
|:---:|:---:|:---:|:---:|
| Regressions | 0.55%, 15.82% | 4.97% | 47 |
| Improvements | -25.86%, -0.40% | -6.29% | 111 |
| All | -25.86%, 15.82% | -2.94% | 158 |

# Primary benchmarks
| Benchmark | Profile | Scenario | % Change | Significance Factor |
|:---:|:---:|:---:|:---:|:---:|
| regex-1.5.5 | clippy | incr-patched: byte frequencies | 15.21% | 76.03x
| libc-0.2.124 | clippy | incr-patched: clone | -14.72% | 73.61x
| clap-3.1.6 | clippy | incr-unchanged | -13.43% | 67.14x
| regex-1.5.5 | clippy | incr-patched: reverse | 13.08% | 65.40x
| regex-1.5.5 | clippy | incr-unchanged | -12.10% | 60.50x
| bitmaps-3.1.0 | clippy | incr-unchanged | 11.88% | 59.39x
| unicode-normalization-0.1.19 | clippy | incr-patched: println | -11.16% | 55.79x
| serde-1.0.136 | clippy | incr-patched: println | 10.68% | 53.42x
| syn-1.0.89 | clippy | incr-unchanged | -9.70% | 48.51x
| exa-0.10.1 | clippy | incr-unchanged | -9.33% | 46.66x
| image-0.24.1 | clippy | incr-patched: println | -9.24% | 46.21x
| hyper-0.14.18 | clippy | incr-full | -8.73% | 43.67x
| cranelift-codegen-0.82.1 | clippy | incr-full | -8.64% | 43.20x
| diesel-1.4.8 | clippy | incr-full | 8.17% | 40.84x
| cargo-0.60.0 | clippy | incr-unchanged | -7.94% | 39.68x
| clap-3.1.6 | clippy | incr-patched: println | -7.84% | 39.19x
| syn-1.0.89 | clippy | full | -7.59% | 37.93x
| helloworld | clippy | full | 7.52% | 37.62x
| html5ever-0.26.0 | clippy | incr-unchanged | -7.50% | 37.52x
| regex-1.5.5 | clippy | incr-patched: is valid cap letter | 7.35% | 36.73x
| syn-1.0.89 | clippy | incr-patched: println | 7.18% | 35.91x
| unicode-normalization-0.1.19 | clippy | incr-full | -7.12% | 35.62x
| bitmaps-3.1.0 | clippy | full | 7.01% | 35.05x
| helloworld | clippy | incr-full | -6.99% | 34.97x
| libc-0.2.124 | clippy | incr-full | -6.94% | 34.71x
| html5ever-0.26.0 | clippy | incr-patched: println | -6.88% | 34.38x
| regex-1.5.5 | clippy | incr-patched: Compiler new | -6.79% | 33.97x
| ripgrep-13.0.0 | clippy | full | -6.78% | 33.88x
| typenum-1.17.0 | clippy | full | -6.60% | 33.02x
| serde-1.0.136 | clippy | incr-unchanged | 6.21% | 31.03x
| clap-3.1.6 | clippy | incr-full | -6.17% | 30.83x
| hyper-0.14.18 | clippy | incr-unchanged | -6.03% | 30.15x
| html5ever-0.26.0 | clippy | incr-full | -6.03% | 30.13x
| clap-3.1.6 | clippy | full | -5.92% | 29.60x
| diesel-1.4.8 | clippy | incr-patched: println | 5.41% | 27.06x
| cranelift-codegen-0.82.1 | clippy | full | -5.00% | 24.99x
| cranelift-codegen-0.82.1 | clippy | incr-patched: println | 4.98% | 24.90x
| exa-0.10.1 | clippy | incr-full | -4.81% | 24.07x
| regex-1.5.5 | clippy | incr-full | 4.76% | 23.78x
| regex-1.5.5 | clippy | full | -4.72% | 23.59x
| typenum-1.17.0 | clippy | incr-patched: add fn | -4.68% | 23.40x
| unicode-normalization-0.1.19 | clippy | incr-unchanged | 4.65% | 23.26x
| stm32f4-0.14.0 | clippy | incr-unchanged | -4.40% | 21.98x
| typenum-1.17.0 | clippy | incr-full | -4.16% | 20.79x
| serde_derive-1.0.136 | clippy | incr-patched: println | 4.02% | 20.12x
| cranelift-codegen-0.82.1 | clippy | incr-unchanged | 3.85% | 19.23x
| webrender-2022 | clippy | incr-patched: println | 3.74% | 18.71x
| webrender-2022 | clippy | incr-unchanged | 3.51% | 17.53x
| typenum-1.17.0 | clippy | incr-unchanged | 3.46% | 17.30x
| image-0.24.1 | clippy | full | -3.35% | 16.75x
| libc-0.2.124 | clippy | full | -3.25% | 16.26x
| unicode-normalization-0.1.19 | clippy | full | 3.21% | 16.03x
| stm32f4-0.14.0 | clippy | incr-full | -3.01% | 15.03x
| diesel-1.4.8 | clippy | incr-unchanged | 2.99% | 14.94x
| helloworld | clippy | incr-unchanged | -2.97% | 14.87x
| cargo-0.60.0 | clippy | incr-patched: println | -2.96% | 14.81x
| exa-0.10.1 | clippy | incr-patched: printlns | -2.88% | 14.42x
| diesel-1.4.8 | clippy | full | 2.87% | 14.35x
| regex-1.5.5 | clippy | incr-patched: Job | -2.75% | 13.75x
| serde-1.0.136 | clippy | full | -2.68% | 13.41x
| exa-0.10.1 | clippy | full | -2.63% | 13.14x
| ripgrep-13.0.0 | clippy | incr-patched: println | -2.59% | 12.93x
| bitmaps-3.1.0 | clippy | incr-full | -2.53% | 12.65x
| syn-1.0.89 | clippy | incr-full | -2.31% | 11.54x
| image-0.24.1 | clippy | incr-unchanged | 2.20% | 10.99x
| serde_derive-1.0.136 | clippy | incr-unchanged | 2.14% | 10.71x
| html5ever-0.26.0 | clippy | full | 1.86% | 9.28x
| serde-1.0.136 | clippy | incr-full | -1.86% | 9.28x
| libc-0.2.124 | clippy | incr-unchanged | -1.76% | 8.78x
| hyper-0.14.18 | clippy | full | -1.74% | 8.70x
| bitmaps-3.1.0 | clippy | incr-patched: println | 1.69% | 8.45x
| image-0.24.1 | clippy | incr-full | 1.54% | 7.70x
| ripgrep-13.0.0 | clippy | incr-full | -1.40% | 7.01x
| regex-1.5.5 | clippy | incr-patched: println | -1.10% | 5.48x
| webrender-2022 | clippy | full | -1.06% | 5.32x
| serde_derive-1.0.136 | clippy | incr-full | -0.95% | 4.75x
| serde_derive-1.0.136 | clippy | full | -0.94% | 4.72x
| cargo-0.60.0 | clippy | full | 0.87% | 4.33x
| ripgrep-13.0.0 | clippy | incr-unchanged | -0.84% | 4.18x
| webrender-2022 | clippy | incr-full | -0.71% | 3.56x
| helloworld | clippy | incr-patched: println | -0.65% | 3.25x
| hyper-0.14.18 | clippy | incr-patched: println | -0.40% | 2.01x

# Secondary benchmarks
| Benchmark | Profile | Scenario | % Change | Significance Factor |
|:---:|:---:|:---:|:---:|:---:|
| wf-projection-stress-65510 | clippy | incr-unchanged | -25.86% | 129.30x
| issue-46449 | clippy | incr-unchanged | -19.10% | 95.51x
| externs | clippy | full | -16.99% | 84.96x
| issue-46449 | clippy | incr-patched: empty 3072 | -16.75% | 83.73x
| token-stream-stress | clippy | incr-unchanged | 15.82% | 79.10x
| await-call-tree | clippy | incr-full | 15.63% | 78.14x
| deep-vector | clippy | incr-full | -14.98% | 74.90x
| coercions | clippy | full | -14.94% | 74.71x
| unused-warnings | clippy | full | -14.86% | 74.30x
| coercions | clippy | incr-patched: add static arr item | -14.65% | 73.27x
| deep-vector | clippy | incr-unchanged | -14.36% | 71.78x
| unify-linearly | clippy | incr-full | -13.16% | 65.79x
| ucd | clippy | incr-full | -13.07% | 65.37x
| issue-58319 | clippy | incr-full | -12.50% | 62.50x
| unused-warnings | clippy | incr-unchanged | -11.67% | 58.37x
| issue-58319 | clippy | full | -11.17% | 55.87x
| issue-46449 | clippy | full | -10.69% | 53.46x
| externs | clippy | incr-unchanged | -10.15% | 50.77x
| many-assoc-items | clippy | incr-unchanged | -10.07% | 50.33x
| issue-88862 | clippy | incr-unchanged | -9.61% | 48.06x
| issue-46449 | clippy | incr-patched: u32 3072 | -8.88% | 44.39x
| issue-46449 | clippy | incr-patched: io error 6144 | -8.80% | 44.01x
| unify-linearly | clippy | full | -8.35% | 41.76x
| await-call-tree | clippy | incr-unchanged | 7.91% | 39.53x
| wg-grammar | clippy | incr-unchanged | -7.32% | 36.59x
| externs | clippy | incr-full | 6.82% | 34.09x
| deep-vector | clippy | incr-patched: println | -6.52% | 32.61x
| projection-caching | clippy | incr-full | -6.40% | 32.01x
| ripgrep-13.0.0-tiny | clippy | full | -6.39% | 31.97x
| wg-grammar | clippy | full | -6.23% | 31.14x
| issue-46449 | clippy | incr-patched: static str 6144 | -6.15% | 30.75x
| deep-vector | clippy | full | -5.97% | 29.86x
| match-stress | clippy | incr-full | -5.85% | 29.25x
| unify-linearly | clippy | incr-patched: dummy fn | -5.42% | 27.11x
| derive | clippy | full | 5.42% | 27.09x
| issue-88862 | clippy | incr-full | -5.36% | 26.81x
| regression-31157 | clippy | incr-unchanged | -5.31% | 26.54x
| regression-31157 | clippy | incr-full | -5.24% | 26.20x
| issue-88862 | clippy | full | -5.24% | 26.18x
| unused-warnings | clippy | incr-full | -4.83% | 24.13x
| regression-31157 | clippy | incr-patched: println | -4.70% | 23.50x
| ucd | clippy | full | -4.35% | 21.76x
| unify-linearly | clippy | incr-unchanged | -4.00% | 19.98x
| derive | clippy | incr-full | 3.91% | 19.53x
| token-stream-stress | clippy | full | 3.87% | 19.35x
| coercions | clippy | incr-full | -3.85% | 19.25x
| many-assoc-items | clippy | incr-full | -3.81% | 19.06x
| regression-31157 | clippy | full | 3.64% | 18.21x
| tuple-stress | clippy | incr-unchanged | -3.64% | 18.18x
| wf-projection-stress-65510 | clippy | full | 3.56% | 17.80x
| wf-projection-stress-65510 | clippy | incr-full | -3.48% | 17.42x
| derive | clippy | incr-unchanged | -3.42% | 17.11x
| helloworld-tiny | clippy | full | -3.34% | 16.68x
| ucd | clippy | incr-unchanged | 3.32% | 16.61x
| ctfe-stress-5 | clippy | incr-full | -3.11% | 15.57x
| many-assoc-items | clippy | full | -3.01% | 15.07x
| match-stress | clippy | incr-unchanged | -2.84% | 14.21x
| coercions | clippy | incr-unchanged | -2.71% | 13.57x
| ctfe-stress-5 | clippy | incr-unchanged | -2.66% | 13.32x
| ctfe-stress-5 | clippy | full | -2.49% | 12.44x
| token-stream-stress | clippy | incr-full | 2.19% | 10.94x
| tt-muncher | clippy | incr-unchanged | 2.01% | 10.03x
| issue-46449 | clippy | incr-patched: u8 3072 | -1.93% | 9.67x
| tuple-stress | clippy | incr-patched: new row | -1.81% | 9.05x
| await-call-tree | clippy | full | -1.81% | 9.04x
| deep-vector | clippy | incr-patched: add vec item | 1.57% | 7.85x
| tuple-stress | clippy | incr-full | 1.56% | 7.80x
| issue-58319 | clippy | incr-unchanged | 1.50% | 7.52x
| projection-caching | clippy | incr-unchanged | -1.25% | 6.26x
| unused-warnings | clippy | incr-patched: dummy fn | -0.93% | 4.64x
| wg-grammar | clippy | incr-full | 0.84% | 4.21x
| tt-muncher | clippy | incr-full | 0.66% | 3.28x
| match-stress | clippy | full | 0.63% | 3.13x
| tuple-stress | clippy | full | 0.55% | 2.76x
| coercions | clippy | incr-patched: println | -0.43% | 2.17x
| issue-46449 | clippy | incr-full | -0.42% | 2.12x
</details>


If you're interested in the whole `results.db` file, I can upload it to a gist and share the link.

Reply:
I'm not sure about benchmarks for Clippy, but I'd guess that this is fine for this update. :)